### ISSUE TITLE
read_parquet should fail if a glob pattern results in multiple matches

### ISSUE THREAD

**Issue by bneijt** [pola-rs/polars#5863]
### Problem description

Currently (0.15.7) if I do a `read_parquet` with a glob pattern (with an s3:// prefix, which might be an issue) and the result has multiple matches, `read_parquet` will simply read only the first match (whatever that might be, not sure how this is selected).

This makes it really easy to read only a single parquet file while you meant to read multiple (`part-0.parquet`, `part-1.parquet` using with `/*.parquet`). However, you only get the result of a single file!

I would like to have polars throw an error if I get data from a single file from a glob pattern when I have multiple matches to make sure I don't load partial data by accident.

**Comment by ritchie46** [comment]
I think this is related to fsspec. Locally we read all matches of the glob pattern.

**Comment by JMrziglod** [comment]
It looks like `fsspec.open` returns only the first file when getting a glob pattern (path with \*). The proper command which can handle glob patterns properly would be `fsspec.open_files` which returns a list of files instead. 

I think (without testing it!) the problem might be in this [if-condition](https://github.com/pola-rs/polars/blob/1f547e50d26065085cd3e197871307718447c796/py-polars/polars/internals/io.py#L143) in `polars.internals.io._prepare_file_args`. It should exclude glob patterns in the filename (e.g. `"*" not in file`) and the next [if-condition](https://github.com/pola-rs/polars/blob/1f547e50d26065085cd3e197871307718447c796/py-polars/polars/internals/io.py#L158) should then deal with them instead.

If you want, I can open a PR for this.

**Comment by tocab** [comment]
I also run into that problem. I have created a colab notebook that highlights it and compares the expected behavior with dask: 
https://colab.research.google.com/drive/1vosWKSZlBun9W8yI9pO3QI5Mww6HNSHi?usp=sharing

I hope that helps with debugging the problem.

**Comment by legout** [comment]
Same problem here. Current workaround is using pyarrow dataset. But I would prefer a polars native solution.

**Comment by legout** [comment]
> It looks like `fsspec.open` returns only the first file when getting a glob pattern (path with *). The proper command which can handle glob patterns would be `fsspec.open_files` which returns a list of files instead.
> 
> I think (without testing it!) the problem might be in this [if-condition](https://github.com/pola-rs/polars/blob/1f547e50d26065085cd3e197871307718447c796/py-polars/polars/internals/io.py#L143) in `polars.internals.io._prepare_file_args`. It should exclude glob patterns in the filename (e.g. `"*" not in file`) and the next [if-condition](https://github.com/pola-rs/polars/blob/1f547e50d26065085cd3e197871307718447c796/py-polars/polars/internals/io.py#L158) should then deal with them instead.
> 
> If you want, I can open a PR for this.

It seems, that there is no `open_files` in s3fs (fsspec s3 implementation), which is used by polars.

**Comment by bneijt** [comment]
I ran into this issue when dealing with a hive dataset and ended up [publishing a one-class package](https://pypi.org/project/polario/) with some of the work-around code to deal hive flavor partitioning in polars.

Best approach to iterate through partitions _I could find_ was using the [pyarrow dataset fragment expressions](https://github.com/bneijt/polario/blob/main/polario/dataset.py#L102).

Another approach is to use the fsspec ls on top of the s3 driver and use the result to load those files into a dataframe, but for a Hive flavor dataset you will have to make sure you add the [partition column values after loading them](https://github.com/bneijt/polario/blob/main/polario/dataset.py#L126):
```
fs = fsspec.filesystem(self.location.scheme)
parquet_files = [
    fragment_location
    for fragment_location in fs.ls(partition_location, detail=False)
    if fragment_location.endswith(".parquet")
]
```

**Comment by PaulRudin** [comment]
I hit this too. FWIW - I think this is probably a bug rather than an enhancement...

**Comment by cjackal** [comment]
Closed by #10098
